include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -168,39 +318,112 @@ class __sort_global_kernel;
+        using _CustomName = oneapi::dpl::__internal::__policy_kernel_name<_ExecutionPolicy>;
+        using _LeafDPWI = std::integral_constant<std::uint16_t, _LeafSorterT::__data_per_workitem>;
+        // TODO: split the submitter into multiple ones to avoid extra compilation of kernels
+        // - _LeafSortKernel does not need _IndexT
+        // - _GlobalSortKernel does not need _LeafDPWI
+        // - _CopyBackKernel does not need either of them
+        using _LeafSortKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+            __sort_leaf_kernel<_CustomName, _IndexT, _LeafDPWI>>;
+        using _GlobalSortKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+            __sort_global_kernel<_CustomName, _IndexT, _LeafDPWI>>;
+        using _CopyBackKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+            __sort_copy_back_kernel<_CustomName, _IndexT, _LeafDPWI>>;
+
+        return __parallel_sort_submitter<_IndexT, _LeafSortKernel, _GlobalSortKernel, _CopyBackKernel>()(
+            std::forward<_ExecutionPolicy>(__exec), std::forward<_Range>(__rng), __comp, __leaf_sorter);
+    }
+};
+
+template <typename _IndexT, typename _ExecutionPolicy, typename _Range, typename _Compare>
+auto
+__submit_selecting_leaf(_ExecutionPolicy&& __exec, _Range&& __rng, _Compare __comp)
+{
+    // 8 is the maximum reasonable value for bubble sub-group sorter due to algorithm complexity
+    // TODO: reconsider the value if another algorithm is used,
+    //       or an internal cap is set (e.g. sorting 2 sequences of 4/8 items each)
+    using _Leaf8 = __leaf_sorter<8, std::decay_t<_Range>, _Compare>;
+    using _Leaf4 = __leaf_sorter<4, std::decay_t<_Range>, _Compare>;
+    // 2 is the smallest reasonable value for merge-path group sorter since it loads 2 values at least
+    using _Leaf2 = __leaf_sorter<2, std::decay_t<_Range>, _Compare>;
+
+    using _Tp = oneapi::dpl::__internal::__value_t<_Range>;
+
+    const std::size_t __n = __rng.size();
+    auto __device = __exec.queue().get_device();
+
+    const std::size_t __max_wg_size = __device.template get_info<sycl::info::device::max_work_group_size>();
+    std::size_t __wg_size = oneapi::dpl::__internal::__dpl_bit_floor(__max_wg_size);
+
+    const bool __is_cpu = __device.is_cpu();
+    std::uint32_t __max_sg_size{};
+    if (__is_cpu)
+    {
+        const auto __sg_sizes = __device.template get_info<sycl::info::device::sub_group_sizes>();
+        __max_sg_size = *std::max_element(__sg_sizes.begin(), __sg_sizes.end());
+    }
+    // Assume CPUs handle one sub-group (SIMD) per CU;
+    // Assume GPUs handle multiple sub-groups per CU,
+    // while the maximum work-group size takes hardware multithreading (occupancy) into account
+    const std::size_t __max_hw_wg_size = __is_cpu ? __max_sg_size : __max_wg_size;
+    const auto __max_cu = __device.template get_info<sycl::info::device::max_compute_units>();
```

Comment:
I decided to stick to that value because it provides significantly better performance comparing to the previous leaf implementation, but retains generalization.

Let me share calculations regarding Intel® Data Center GPU Max (1100 model, for example) saturation:

1.  Real saturation (how it maps HW in my view): 
`VE Count * Max Sub-Group Size * Max Occupancy = 448 * 32 * 8 = 114'688`
2. Current calculation: 
`VE Count * Max Work-Group Size = 448 * 1024 = 458'752`
3. If CU count was reported as XC count: 
`XC Count * Max Work-Group Size = 57'344`

Note that there is ticket for Intel GPU driver to rise max work-group size to 2048 to achieve full utilization of XC with a single group. So 1st or 3rd option is what is needed for Intel GPUs. 3rd is better, because it is aligned with what we get for other vendors. So, I suppose saturation is already correct for Nvidia GPUs. 

Currently, the difference between desired and real saturation is 4x for Intel GPU. Let me check the performance effect on dividing it by 4. I did some experiments previously, performance was not affected much, but let me do a more targeted one. 

I will refrain from any vendor dispatch as possible, though. If the performance impact is small, I will leave it as is.

Anyway, I will add a note in our code, and try to submit an issue for reporting XC instead of VE (probably, it was discussed before and there is a reason, but let me try).

Reply:
Thanks, I agree we should avoid vendor dispatch as much as possible and only do it if the performance benefit is large. After this experiment on the performance impact, I am ready to approve.

Reply:
Speed up for different saturation points on Intel® Data Center GPU Max 1550 (1T, 512 VE) against **512 * 1024** saturation point:

  | n | 64 * 1024 | 128 *   1024 | 256 * 1024
-- | -- | -- | -- | --
merge sort (float) | 256 | 0.98 | 1.02 | 1.02
merge sort (float) | 512 | 1.00 | 1.04 | 1.00
merge sort (float) | 1024 | 1.00 | 1.00 | 1.00
merge sort (float) | 2048 | 1.00 | 1.00 | 1.03
merge sort (float) | 4096 | 1.00 | 1.02 | 1.00
merge sort (float) | 8192 | 1.02 | 1.00 | 1.02
merge sort (float) | 16384 | 1.00 | 0.98 | 1.00
merge sort (float) | 32768 | 1.02 | 1.00 | 1.00
merge sort (float) | 65536 | 1.03 | 1.01 | 1.03
merge sort (float) | 131072 | 0.99 | 0.99 | 0.99
merge sort (float) | 262144 | 0.98 | 0.99 | 1.00
merge sort (float) | 524288 | 1.04 | 1.11 | 1.00
merge sort (float) | 1048576 | 1.29 | 1.29 | 1.18
merge sort (float) | 2097152 | 1.15 | 1.14 | 1.15
merge sort (float) | 4194304 | 1.00 | 1.00 | 1.00
merge sort (float) | 8388608 | 1.00 | 1.00 | 1.00
merge sort (float) | 16777216 | 1.00 | 1.00 | 1.00
merge sort (float) | 33554432 | 1.01 | 1.01 | 1.01
merge sort (float) | 67108864 | 1.01 | 1.01 | 1.01
merge sort (float) | 134217728 | 1.01 | 1.01 | 1.01
merge sort (float) | 268435456 | 1.01 | 1.01 | 1.01

I've measured performance adjusting the current saturation point. It appears 4x adjustment to match theoretical HW saturation makes a lot for 500K - 4M sizes.

Reply:
I think this benefit worth thinking about. Let me check, what can be done...

Reply:
I haven't managed to find a solution, which is not controversial, and I've added comments.
Let's proceed with this PR, and discuss this issue more separately.